### PR TITLE
feat: subprocess provider mixin class

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

fixes: #518 

### How I did it

Mostly moved logic from `ape-hardhat` to here; here is the matching `ape-hardhat` PR: https://github.com/ApeWorX/ape-hardhat/pull/30.

Now, plugin developers will need less code to create providers based on subprocesses. So we can probably create a `ganache` one easier now. I am using this elsewhere too, as you may know!

### How to verify it

I tested out both hardhat providers and and ran the `ape-hardhat` test suite locally.
You can do something similar! Try using both hardhat provider in a script or the console or tests. Make sure it works as before.

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: src/ape/api/providers.py [hunk 1]
```diff
@@ -719,3 +737,188 @@ def connection_str(self) -> str:
...
+    """
+
+    PROCESS_WAIT_TIMEOUT = 15
+    process: Optional[Popen] = None
+    is_stopping: bool = False
+
+    @property
+    @abstractmethod
+    def process_name(self) -> str:
+        """The name of the process, such as ``Hardhat node``."""
```
Comment (fubuloubu) [inline src/ape/api/providers.py:754]: Is this note correct?

Comment (fubuloubu) [reply src/ape/api/providers.py:754]: Or wait, is this just used for display purposes? or calling subprocess?

**Review by fubuloubu — APPROVED** [review]

File: src/ape/api/providers.py (hunk 1 above)
Comment (unparalleled-js) [reply src/ape/api/providers.py:754]: Yes this is just for display purposes, like the log statements `"Starting blank"`.

Perhaps there is a better way? hmmmm

The actual command comes from the abstract method `build_command()`.
Both `Hardhat` providers use this hook as needed, like the `mainnet-fork` provider includes the `--fork` and `--block-number` flags.

Comment (unparalleled-js) [reply src/ape/api/providers.py:754]: Oh, and the first command in `ape-hardhat` is the `npx` bin path, so I can't find a reliable way to know the display name of a process based on its command list

Comment (fubuloubu) [reply src/ape/api/providers.py:754]: Okay, was confused for a sec lol

**Pull request merged**